Add config for rasterized symbols cache

rafatower opened the pull request:
What this does:
- Adds a member `bool rasterized_symbols_cache_disabled_;` to the `agg_renderer`
- That flag is initialized in the `setup` of the renderer, through the parameters of the map (at instantiation time).
- That flag is in turn passed to the `agg_markers_renderer_context` at instantiation time.
- The context can take decisions about using/not using the cache based on that flag at render time (`render_marker`).

How the parameter is (or can be) passed:
The map admits extra parameters in the XML. You can see an example here:

<details>

```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE Map[]>
<Map srs="+init=epsg:3857" maximum-extent="-20037508.3,-20037508.3,20037508.3,20037508.3">

<Parameters>
  <Parameter name="maximum-extent"><![CDATA[-20037508.3,-20037508.3,20037508.3,20037508.3]]></Parameter>
  <Parameter name="format">png8:m=h</Parameter>
  <Parameter name="rasterized_symbols_cache_disabled">true</Parameter>
  <Parameter name="interactivity_layer">layer0</Parameter>
  <Parameter name="interactivity_fields">cartodb_id</Parameter>
</Parameters>


<Style name="layer0" filter-mode="first">
  <Rule>
    <MarkersSymbolizer width="7" fill="#4dee83" fill-opacity="0.9" allow-overlap="true" stroke-width="1" stroke="#ffffff" stroke-opacity="1" clip="true" />
  </Rule>
</Style>
<Layer name="layer0"
  cache-features="true"
  srs="+init=epsg:3857">
    <StyleName>layer0</StyleName>
    <Datasource>
       <Parameter name="type"><![CDATA[postgis]]></Parameter>
       <Parameter name="user"><![CDATA[publicuser]]></Parameter>
       <Parameter name="password"><![CDATA[public]]></Parameter>
       <Parameter name="host"><![CDATA[localhost]]></Parameter>
       <Parameter name="port"><![CDATA[6432]]></Parameter>
       <Parameter name="extent"><![CDATA[-20037508.3,-20037508.3,20037508.3,20037508.3]]></Parameter>
       <Parameter name="row_limit"><![CDATA[0]]></Parameter>
       <Parameter name="simplify_geometries"><![CDATA[false]]></Parameter>
       <Parameter name="use_overviews"><![CDATA[true]]></Parameter>
       <Parameter name="persist_connection"><![CDATA[false]]></Parameter>
       <Parameter name="max_size"><![CDATA[500]]></Parameter>
       <Parameter name="twkb_encoding"><![CDATA[true]]></Parameter>
       <Parameter name="srid"><![CDATA[3857]]></Parameter>
       <Parameter name="table"><![CDATA[(select * from ne_10m_populated_places_simple) as cdbq]]></Parameter>
       <Parameter name="dbname"><![CDATA[cartodb_dev_user_5659debf-7f98-4403-9c63-be961436eda9_db]]></Parameter>
       <Parameter name="geometry_field"><![CDATA[the_geom_webmercator]]></Parameter>
    </Datasource>
  </Layer>

</Map>
```

</details>

Those parameters can be injected from Windshaft (`lib/windshaft/renderers/mapnik/factory.js`) and they can be mapped to map parameters by grainstore: <details>
```patch
$ git diff lib/grainstore/mml-builder/mml-builder-inline.js
diff --git a/lib/grainstore/mml-builder/mml-builder-inline.js b/lib/grainstore/mml-builder/mml-builder-inline.js
index 99f3bca..1efbd0c 100644
--- a/lib/grainstore/mml-builder/mml-builder-inline.js
+++ b/lib/grainstore/mml-builder/mml-builder-inline.js
@@ -119,6 +119,9 @@ function MMLBuilderInline(params, options) {
     if ( options.mapnik_tile_format ) {
         this.grainstore_map.format = options.mapnik_tile_format;
     }
+    if ( params.rasterized_symbols_cache_disabled !== undefined ) {
+        this.grainstore_map.rasterized_symbols_cache_disabled = params.rasterized_symbols_cache_disabled
+    }
 
     this.interactivity = params.interactivity;
     if ( _.isString(this.interactivity) ) {
@@ -242,6 +245,8 @@ MMLBuilderInline.prototype.baseMML = function() {
     mml['maximum-extent'] = this.grainstore_map['maximum-extent'];
     mml.format = this.grainstore_map.format || 'png';
     mml.Layer = [];
+    mml.rasterized_symbols_cache_disabled = this.grainstore_map.rasterized_symbols_cache_disabled;
 
     for (var i=0; i<tables.length; ++i) {
         var table = tables[i];
```
</details>

Commit:
Config to enable/disable cache at layer level #40

Add rasterized_symbols_cache_enabled config param at the layer level, to
enable or disable that cache. It is enabled by default.

Commit:
Read a config value from vars #40

Read the config value from vars, and use the default value defined for
the symbolizers.

It doesn't strike me as the best option, but I'm struggling to have an
option first, then iterate. The layer code is to be reviewed later on.

Commit:
Move config to markers_dispatch_params #40

Commit:
Remove rasterized_symbols_cache_enabled from layers #40

Commit:
rasterized_symbols_cache_enabled now defaults to false #40

Commit:
Change the var and logic s/enabled/disabled/ #40

Commit:
Working version (WIP) #40

This needs a lot of clean up, to leave just what's needed.

Commit:
Remove uneeded code #40

Remove code that's not needed from `vector_markers_dispatch` and
`markers_dispatch_params`.

Commit:
Remove uneeded code from symbolizer_keys #40

Commit:
Make sure bool member is initialized #40

Make sure rasterized_symbols_cache_disabled_ is initialized based on the
map parameters.

Commit:
Simplify the agg_renderer code #40

Algunenano commented:
A couple of things:
* I'd name it `markers_symbolizer_cache_disabled_` instead of `rasterized_symbols_cache_disabled_` since it's more similar to  what it's doing and more similar to the name of the affected files.
* You are missing the cache in `mapnik/src/renderer_common/render_markers_symbolizer.cpp` which I think should be covered by this flag too.

What I've done for the metrics is to add them in `mapnik/include/mapnik/renderer_common/render_markers_symbolizer.hpp` so it's available for all renderers and you can use them in the `render_markers_symbolizer.cpp` too (via `renderer_context_.metrics_`). This forces to edit more code (since the constructors need to be redefined) but allows for more flexibility.

I foresee a painful merge since we are editing the same files.

rafatower commented:
First of all, thanks for the prompt comments.

I updated the description, at least to keep a record of what I'm doing.

> I'd name it `markers_symbolizer_cache_disabled_` instead of `rasterized_symbols_cache_disabled_` since it's more similar to what it's doing and more similar to the name of the affected files.

It's a matter of semantics. The name shall refer to what's cached or to who is the owner/user of it? anyway, whatever is clear among us is fine to me :+1: 

> You are missing the cache in `mapnik/src/renderer_common/render_markers_symbolizer.cpp` which I think should be covered by this flag too.

Yep, I knew I forgot about some cache :sweat_smile: , thanks for the reminder!

> I foresee a painful merge since we are editing the same files.

That's why I wanted to share with you, to be ahead and agree on the approach, cause it's not only mapnik but the reverse dependencies and... you know. PS: any suggestion about how to improve the situation I'm all ears. I mean, we can split tasks differently, if that helps.

Commit:
Rename to markers_symbolizer_caches_disabled #40

```
s/rasterized_symbols_cache_disabled/markers_symbolizer_caches_disabled/g
```

Commit:
Move config flag to generic context #40

Commit:
Merge remote-tracking branch 'carto/v3.0.15-carto' into 40-add-config-rasterized_symbols_cache

Commit:
Merge remote-tracking branch 'carto/v3.0.15-carto' into 40-add-config-rasterized_symbols_cache

Commit:
Deal with cached_attributes config flag #40

Commit:
Deal with cached_ellipses config flag #40

Commit:
Merge remote-tracking branch 'carto/v3.0.15-carto' into 40-add-config-rasterized_symbols_cache

Conflicts:
	include/mapnik/renderer_common/render_markers_symbolizer.hpp
	src/agg/process_markers_symbolizer.cpp

Commit:
Update changelog w/ markers_symbolizer_caches_disabled

Commit:
Fix: marker_ptr is only searched and stored when using the cache

Otherwise the marker_ptr is used from the so called
stock_vector_marker.
